addon_service/addon_operation_invocation/models.py

```diff
@@ -66,13 +85,13 @@ def owner_uri(self) -> str:
 
     @property
     def imp_cls(self) -> type[AddonImp]:
-        return self.thru_account.imp_cls
+        return self.account.imp_cls
 
     @property
-    def storage_imp_config(self) -> StorageConfig:
+    def config(self) -> StorageConfig | CitationConfig:
         if self.thru_addon:
-            return self.thru_addon.storage_imp_config
-        return self.thru_account.storage_imp_config
+            return self.addon.config
+        return self.account.config
```

Comment:
getting interface-specific config seems best kept in separate interface-specific functions (which might easily/inherently validate that the account/addon is the expected type)

also (less importantly), that logic might make more sense somewhere else, since `AddonOperationInvocation` itself doesn't otherwise need to know whether the addon does storage or citations or whatever -- looks like `addon_service.addon_imp.instantiation` and `addon_service.tasks.invocation` haven't been re-thought since adding citations, could be redone to more cleanly handle interface-specific concerns...

Reply:
I think that discerning whether it is citation or storage invocation based on whether we have `storage_imp_config` or `citation_imp_config` is a crutch. 
Moreover having separate property name for separate addon types causes issues during instantiation and requires different factories.
I do agree that we have to rethink `AddonOperationInvocation` now that we have citations and storage in the picture.

Reply:
hmm, yeah that makes sense -- i'll admit i was assuming different factories per addon type/interface (looking at the [get_citation_addon_instance stub](https://github.com/CenterForOpenScience/gravyvalet/blob/81703b6dd59720e24339925e7310b42a04fab472/addon_service/addon_imp/instantiation.py#L60-L65)), so type-specific config properties would not be used to discern type, but used by code that already knows which type... but all that is irrelevant if we can consolidate imp-stantiation in a reasonable way

i think the code smell i was actually responding to is the return type with ever-growing union `def config(self) -> ZConfig | YConfig | XConfig | ...` -- would be better if there were a reasonable `AddonImpConfig` base class (is there anything logically shared across all possible addon types?)

Reply:
I get your point, but at least we get quickly exhaustive list (as we have only storage, citation and compute addons)
Moreover, we can narrow the type of config in subclasses of our model. I may know similar fields for citation and storage configs, but comput is a dark horse for now. If you still think this is an issue, I'll try to fix it.